### PR TITLE
fix reconcile too long and some error in log

### PR DESCRIPTION
**Pull request by ldpliu**
https://issues.redhat.com/browse/ACM-7913
1. When Kafka or Postgre is starting, we should not return errors in reconcile. We just need to wait it started
2. We should not return `mgh is not exist` errors before `mgh` is created. just need to wait it created.

### CODE REVIEW

**Comment by clyang82** [comment]
Could you append the root cause why the reconciler takes too long in the description? Thanks.

**Comment by clyang82** [comment]
It may be related with `DefaultControllerRateLimiter` https://github.com/kubernetes-sigs/controller-runtime/blob/15d79283523517f52ee30b28d2ae4c1ea3f113da/pkg/controller/controller.go#L146. 

NewItemExponentialFailureRateLimiter(5*time.Millisecond, 1000*time.Second), the maxDelay is 1000 second. that can be a root cause. I would like to have a try with customized option in controller manager https://github.com/stolostron/multicluster-global-hub/blob/7da17d464be9b1db46702f1790200245c7ae99ab/operator/pkg/controllers/hubofhubs/globalhub_controller.go#L482 with the following code:

```
WithOptions(controller.Options{
			RateLimiter: workqueue.NewItemFastSlowRateLimiter(5*time.Millisecond, time.Second*5, 5),
		}).
```

**Comment by ldpliu** [comment]
> It may be related with `DefaultControllerRateLimiter` https://github.com/kubernetes-sigs/controller-runtime/blob/15d79283523517f52ee30b28d2ae4c1ea3f113da/pkg/controller/controller.go#L146.
> 
> NewItemExponentialFailureRateLimiter(5_time.Millisecond, 1000_time.Second), the maxDelay is 1000 second. that can be a root cause. I would like to have a try with customized option in controller manager
> 
> https://github.com/stolostron/multicluster-global-hub/blob/7da17d464be9b1db46702f1790200245c7ae99ab/operator/pkg/controllers/hubofhubs/globalhub_controller.go#L482
> with the following code:
> 
> ```
> WithOptions(controller.Options{
> 			RateLimiter: workqueue.NewItemFastSlowRateLimiter(5*time.Millisecond, time.Second*5, 5),
> 		}).
> ```

Yes, I append some details in description.
But I think the root cause is we returned too much errors when we wait `kafka or postgre` starting. We just need to wait they started instead of always return errors .

File: operator/pkg/utils/utils.go [hunk 1]
```diff
@@ -252,3 +257,37 @@ func CopyMap(newMap, originalMap map[string]string) {
 		newMap[key] = value
 	}
 }
+
+func WaitGlobalHubReady(ctx context.Context,
+	client client.Client,
+	interval time.Duration) (*globalhubv1alpha4.MulticlusterGlobalHub, error) {
+	mghInstance := &globalhubv1alpha4.MulticlusterGlobalHub{}
+
+	if err := wait.PollInfinite(interval, func() (bool, error) {
```
[the remark this answers is not in the document]
Comment (ldpliu) [reply operator/pkg/utils/utils.go]: > Should we anticipate the check running indefinitely if certain pods are not ready? Perhaps adding a timeout would be a better approach.

Done

File: operator/pkg/utils/utils.go [hunk 2]
```diff
@@ -252,3 +257,37 @@ func CopyMap(newMap, originalMap map[string]string) {
...
+		mghInstance = &mghList.Items[0]
+		if !mghList.Items[0].DeletionTimestamp.IsZero() {
+			klog.Info("MulticlusterGlobalHub is deleting")
+			return false, nil
+		}
+
+		if meta.IsStatusConditionTrue(mghInstance.Status.Conditions, condition.CONDITION_TYPE_GLOBALHUB_READY) {
+			return true, nil
+		}
+		klog.Info("Wait MulticlusterGlobalHub is ready")
```
Comment (clyang82) [inline operator/pkg/utils/utils.go:287]: maybe just print the logs once at the top of this method to tell user we are waiting for mcgh ready. It is too noisy to print every 10 seconds.

File: operator/pkg/controllers/hubofhubs/globalhub_kafka.go [hunk 3]
```diff
@@ -135,6 +135,9 @@ func (r *MulticlusterGlobalHubReconciler) WaitForKafkaClusterReady(ctx context.C
 		Namespace: config.GetDefaultNamespace(),
 	}, kafkaCluster)
 	if err != nil {
+		r.Log.Error(err, "Failed to get Kafka cluster",
```
Comment (clyang82) [inline operator/pkg/controllers/hubofhubs/globalhub_kafka.go:138]: Can we add log level here?

File: operator/pkg/controllers/hubofhubs/globalhub_controller.go [hunk 4]
```diff
@@ -131,31 +133,40 @@ type MiddlewareConfig struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
 func (r *MulticlusterGlobalHubReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	if len(req.Namespace) == 0 || len(req.Name) == 0 {
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+	}
```
Comment (clyang82) [inline operator/pkg/controllers/hubofhubs/globalhub_controller.go:138]: what is the case that the req.Name or Namespace can be empty?

File: operator/pkg/condition/condition.go [hunk 5]
```diff
@@ -74,6 +74,14 @@ const (
 	CONDITION_MESSAGE_LEAFHUB_DEPLOY_FAILED = "Leaf Hub Deployed FAILED"
 )
 
+const (
+	CONDITION_TYPE_GLOBALHUB_READY    = "Ready"
+	CONDITION_REASON_GLOBALHUB_READY  = "MulticlusterGlobalHubReady"
+	CONDITION_MESSAGE_GLOBALHUB_READY = "Multicluster Global Hub is ready"
+	CONDITION_REASON_GLOBALHUB_FAILED = "MulticlusterGlobalHubFailed"
+	CONDITION_REASON_GLOBALHUB_PAUSED = "MulticlusterGlobalHubPaused"
+)
```
Comment (clyang82) [inline operator/pkg/condition/condition.go:83]: I like the idea to add `Ready` condition to indicate the global hub is ready for process. Thank you

File: operator/pkg/controllers/addon/addon_controller.go [hunk 6]
```diff
@@ -111,7 +114,10 @@ func (a *HoHAddonController) Start(ctx context.Context) error {
 		ControllerConfig:     a.ControllerConfig,
 		LogLevel:             a.LogLevel,
 	}
-
+	_, err = utils.WaitGlobalHubReady(ctx, a.client, 10*time.Second)
```
Comment (clyang82) [inline operator/pkg/controllers/addon/addon_controller.go:117]: will it too long? can we check it in 5 seconds like what we did in the past?

File: operator/pkg/controllers/hubofhubs/globalhub_controller.go [hunk 7]
```diff
@@ -131,31 +133,40 @@ type MiddlewareConfig struct {
...
 			return ctrl.Result{}, nil
 		}
 		// Error reading the object - requeue the request.
 		r.Log.Error(err, "Failed to get MulticlusterGlobalHub")
 		return ctrl.Result{}, err
 	}
-
 	if config.IsPaused(mgh) {
 		r.Log.Info("mgh reconciliation is paused, nothing more to do")
+		conditionErr := AddFailedCondition(ctx, r.Client, mgh, "mgh reconciliation is paused")
```
Comment (clyang82) [inline operator/pkg/controllers/hubofhubs/globalhub_controller.go:156]: we do not need to handle the pause case. it is only for developer. And `pause` does not mean the condition is failed.

File: operator/pkg/controllers/hubofhubs/globalhub_controller.go [hunk 8]
```diff
@@ -167,12 +178,41 @@ func (r *MulticlusterGlobalHubReconciler) Reconcile(ctx context.Context, req ctr
...
+		return ctrl.Result{}, err
+	}
+	//Watch middleware avaible every 20 seconds
+	if needReconcileMiddleware {
+		r.Log.V(2).Info("ReconcileMiddleware", "message", message)
+		conditionErr := AddFailedCondition(ctx, r.Client, mgh, message)
+		if conditionErr != nil {
+			return ctrl.Result{}, conditionErr
+		}
+		return reconcile.Result{RequeueAfter: 20 * time.Second}, nil
```
Comment (clyang82) [inline operator/pkg/controllers/hubofhubs/globalhub_controller.go:197]: wait for 20 seconds is too long. I would like to keep 5 seconds.

File: operator/pkg/controllers/hubofhubs/globalhub_controller.go [hunk 9]
```diff
@@ -167,12 +178,41 @@ func (r *MulticlusterGlobalHubReconciler) Reconcile(ctx context.Context, req ctr
 		return ctrl.Result{Requeue: true}, err
 	}
 
-	if result, err := r.ReconcileMiddleware(ctx, mgh); err != nil {
-		return result, err
+	needReconcileMiddleware, message, err := r.ReconcileMiddleware(ctx, mgh)
```
Comment (clyang82) [inline operator/pkg/controllers/hubofhubs/globalhub_controller.go:181]: How about keep the method no change, then AddFailedCondition at the beginning of `Reconcile()` after `if config.IsPaused(mgh) {`,
update the Condition is true at the end of `Reconcile`. What do you think?

**Review by clyang82 — APPROVED** [review]
Great enhancement. Thanks @ldpliu

**Pull request merged**